### PR TITLE
Fixes constness in SSL_CTX_set_srp_username and SSL_CTX_set_srp_password functions

### PR DESCRIPTION
**Pull request by flowher**
Currently  SSL_CTX_set_srp_username/password functions take char* argument for username/password value. In an application level code those values are very often const (user provided data). In such cases, when passing those values to OpenSSL library either dirty cast needs to be performed to remove constness, or const value needs to be copied to temporary location (which for SSL_CTX_set_srp_username is useless as this function copies again username value in ssl3_ctx_ctrl function).

In this patch I try to cleanup API, so that const values also can be passed to functions. Valgrind doesn't show any additional memory leaks (I've run test-apps from OpenSSL and also this one: https://github.com/flowher/srp_client_server_implementation_with_openssl/).

### CODE REVIEW

File: ssl/s3_lib.c [hunk 1]
```diff
@@ -4015,6 +4015,8 @@ long ssl3_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_TLS_EXT_SRP_PASSWORD:
         ctx->srp_ctx.SRP_give_srp_client_pwd_callback =
             srp_password_from_info_cb;
+        if(ctx->srp_ctx.info != NULL)
```
Comment (richsalz) [inline ssl/s3_lib.c]: the check for NULL isn't needed.

File: ssl/s3_lib.c [hunk 2]
```diff
@@ -4007,7 +4007,7 @@ long ssl3_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
             SSLerr(SSL_F_SSL3_CTX_CTRL, SSL_R_INVALID_SRP_USERNAME);
             return 0;
         }
-        if ((ctx->srp_ctx.login = BUF_strdup((char *)parg)) == NULL) {
+        if ((ctx->srp_ctx.login = (char *)parg) == NULL) {
```
Comment (ekasper) [inline ssl/s3_lib.c]: The SSL_ctrl is part of the public API and, therefore, you can't really change pointer ownership semantics.

File: ssl/tls_srp.c [hunk 3]
```diff
@@ -499,14 +512,16 @@ char *SSL_get_srp_userinfo(SSL *s)
 # define tls1_ctx_ctrl ssl3_ctx_ctrl
 # define tls1_ctx_callback_ctrl ssl3_ctx_callback_ctrl
 
-int SSL_CTX_set_srp_username(SSL_CTX *ctx, char *name)
+int SSL_CTX_set_srp_username(SSL_CTX *ctx, const char *name)
 {
-    return tls1_ctx_ctrl(ctx, SSL_CTRL_SET_TLS_EXT_SRP_USERNAME, 0, name);
+    char* name_tmp = BUF_strdup(name);
+    return tls1_ctx_ctrl(ctx, SSL_CTRL_SET_TLS_EXT_SRP_USERNAME, 0, name_tmp);
```
Comment (ekasper) [inline ssl/tls_srp.c]: This also leaves the question: who owns name_tmp if ctrl fails. In practice it maybe can't but ownership semantics must be well defined.

This MR needs to be reworked to preserve SSL_ctrl behaviour. You should be able to push the dup into the ctrl I think.

File: ssl/s3_lib.c (hunk 1 above)
Comment (flowher) [inline ssl/s3_lib.c]: I can continue on this patch, but are you interested in merging it in?

**Comment by flowher** [comment]
I can continue on this patch, but are you interested in merging it in? I just want to make sure I'm not going to waste my time.

**Comment by mattcaswell** [comment]
Ping @richsalz @ekasper . What is happening with this?

**Comment by richsalz** [comment]
Looks like there's some feedback that hasn't been addressed.  This still seems not ready to merge.

**Comment by flowher** [comment]
Yep, waiting for your answer since 15th of march.

**Comment by richsalz** [comment]
Two people expressed an interest and pointed out some need for improvement.  Sorry if that wasn't enough to go on.

**Comment by flowher** [comment]
Ok, I'll continue. 
FYI : similar intrest and comments on another PR, but finally it's not going to be merge, obiously double checking is needed - that's why I'm asking.

**Comment by richsalz** [comment]
Once the issues are addressed I will review and plus-one.  Up to @mattcaswell or someone else to do the other review and merge.

**Comment by mattcaswell** [comment]
I'd prefer @ekasper to review as she already started this. But failing that I will look at this.

**Comment by flowher** [comment]
Thanks, 
I'll try to fix it and address comments soon

File: ssl/tls_srp.c [hunk 4]
```diff
@@ -476,14 +489,16 @@ char *SSL_get_srp_userinfo(SSL *s)
...
 }
 
-int SSL_CTX_set_srp_password(SSL_CTX *ctx, char *password)
+int SSL_CTX_set_srp_password(SSL_CTX *ctx, const char *password)
 {
-    return tls1_ctx_ctrl(ctx, SSL_CTRL_SET_TLS_EXT_SRP_PASSWORD, 0, password);
+    char* pass_tmp = BUF_strdup(password);
```
Comment (flowher) [inline ssl/tls_srp.c]: same comment as above

**Pull request closed without merging**